### PR TITLE
Check for invalid mutations in namespaces

### LINKED ISSUE #129: Storage - check for invalid mutations on structs
depends on: https://github.com/Synthetixio/synthetix-v3/issues/156

### PR DESCRIPTION
**Pull request by leomassazza**
fixes #129 
This PR adds some helper utils to create a map of all structs referenced in the project and a comparator for that map to find mutations between previous and current version

### CODE REVIEW

File: packages/core-js/utils/ast.js [hunk 1]
```diff
@@ -92,7 +96,44 @@ function findFunctionSelectors(contractName, asts) {
   return selectors;
 }
 
+function orderContractsStructMap(structs) {
```
Comment (leomassazza) [inline packages/core-js/utils/ast.js:99]: Order the structs by contract and struct name

File: packages/core-js/utils/ast.js [hunk 2]
```diff
@@ -92,7 +96,44 @@ function findFunctionSelectors(contractName, asts) {
...
+        ? -1
+        : a.struct.name > b.struct.name
+          ? 1
+          : a.struct.name < b.struct.name
+            ? -1
+            : 0
+  );
+}
+
+async function buildContractsStructMap(asts) {
```
Comment (leomassazza) [inline packages/core-js/utils/ast.js:113]: this is the struct map that will be compared between versions

File: packages/core-js/utils/comparator.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
+function compareStorageStructs({ previousStructsMap, currentStructsMap }) {
```
Comment (leomassazza) [inline packages/core-js/utils/comparator.js:1]: This function finds the three kind of mutations that can happen in a struct. 
- Additions (valid mutation)
- Removals
- Modifications

File: packages/deployer/subtasks/prepare-deployment.js [hunk 4]
```diff
@@ -60,8 +60,10 @@ subtask(
   };
 
   if (previousFile) {
+    const { sources: previousSources } = getDeploymentExtendedFiles(previousFile);
```
Comment (leomassazza) [inline packages/deployer/subtasks/prepare-deployment.js:63]: Includes also the sources (that contains the ASTs) in previousDeployment

File: packages/deployer/internal/storage-ast-validator.js [hunk 5]
```diff
@@ -2,13 +2,15 @@ const {
...
   findContractStateVariables,
+  buildContractsStructMap,
 } = require('@synthetixio/core-js/utils/ast');
-const logger = require('@synthetixio/core-js/utils/logger');
+const { compareStorageStructs } = require('@synthetixio/core-js/utils/comparator');
 const filterValues = require('filter-values');
 
 class ModuleStorageASTValidator {
-  constructor(asts) {
+  constructor(asts, previousAsts) {
```
Comment (leomassazza) [inline packages/deployer/internal/storage-ast-validator.js:11]: Adds previous asts to compare struct mutations

File: packages/deployer/internal/storage-ast-validator.js [hunk 6]
```diff
@@ -89,12 +91,42 @@ class ModuleStorageASTValidator {
...
+    const { modifications, removals } = compareStorageStructs({
+      previousStructsMap,
+      currentStructsMap,
+    });
+
+    modifications.forEach((m) => {
+      if (!errors.some((e) => e.contract === m.contract && e.struct === m.struct)) {
+        errors.push({
+          msg: `Invalid mutation found in namespace ${m.contract}.${m.struct}`,
+          contract: m.contract,
```
Comment (leomassazza) [inline packages/deployer/internal/storage-ast-validator.js:113]: contract and struct is added here to check for duplicates in the errors

File: packages/deployer/subtasks/validate-storage.js [hunk 7]
```diff
@@ -1,19 +1,20 @@
...
 const logger = require('@synthetixio/core-js/utils/logger');
-const { getAllContractASTs } = require('@synthetixio/core-js/utils/hardhat');
+const { getAllContractASTs, getAllPreviousContractASTs } = require('../utils/deployments');
 const ModuleStorageASTValidator = require('../internal/storage-ast-validator');
 const { SUBTASK_VALIDATE_STORAGE, SUBTASK_CANCEL_DEPLOYMENT } = require('../task-names');
 
 subtask(SUBTASK_VALIDATE_STORAGE).setAction(async (_, hre) => {
   logger.subtitle('Validating module storage usage');
 
-  const asts = await getAllContractASTs(hre);
```
Comment (leomassazza) [inline packages/deployer/subtasks/validate-storage.js:10]: current asts can be retrieved from the hre.deployer data now.

File: packages/core-js/test/utils/comparator.spec.js [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+            type: 'uint256',
+          },
+        ],
+      },
+    },
+  ];
+
+  describe('compareStorageStructs no updates', function () {
+    it('detects no changes when both structMaps are equal', function () {
+      let currentStructsMap = JSON.parse(JSON.stringify(previousStructsMap));
```
Comment (mjlescano) [inline packages/core-js/test/utils/comparator.spec.js:78]: @leomassazza, What do you think of doinga `clone()` util for `JSON.parse(JSON.stringify(previousStructsMap))`?

File: packages/core-js/utils/comparator.js (hunk 3 above)
Comment (ajsantander) [reply packages/core-js/utils/comparator.js:1]: I think this file should go in the `ast.js` util or we should start having an `ast/` folder in the utils with specialized files.

File: packages/deployer/internal/storage-ast-validator.js [hunk 9]
```diff
@@ -89,12 +91,42 @@ class ModuleStorageASTValidator {
...
+    const errors = [];
+
+    if (!this.previousAsts) {
+      return errors;
+    }
+
+    const previousStructsMap = await buildContractsStructMap(this.previousAsts);
+    const currentStructsMap = await buildContractsStructMap(this.asts);
+
+    const { modifications, removals } = compareStorageStructs({
```
Comment (ajsantander) [inline packages/deployer/internal/storage-ast-validator.js:104]: What about additions? These could be disruptive as well. E.g:
```
// :D
struct HappyStruct {
  address thing;
}
```
mutating to
```
// :(
struct HappyStruct {
  uint newSupahValue;
  address thing;
}
```

**Review by ajsantander — APPROVED** [review]
Beautiful work @leomassazza

File: packages/core-js/utils/ast.js [hunk 10]
```diff
@@ -92,7 +96,45 @@ function findFunctionSelectors(contractName, asts) {
...
+  structs.sort((a, b) =>
+    a.contract.name > b.contract.name
+      ? 1
+      : a.contract.name < b.contract.name
+        ? -1
+        : a.struct.name > b.struct.name
+          ? 1
+          : a.struct.name < b.struct.name
+            ? -1
+            : 0
```
Comment (mjlescano) [inline packages/core-js/utils/ast.js:110]: Here looks like that is doing several times the same comparison, e.g.:  it does 2 times `a.contract.name > b.contract.name`

File: packages/deployer/subtasks/cancel-deployment.js [hunk 11]
```diff
@@ -19,6 +19,6 @@ subtask(
 
   await del(toDelete);
 
-  logger.info('Extiting...');
+  logger.info('Exiting...');
```
Comment (mjlescano) [inline packages/deployer/subtasks/cancel-deployment.js:22]: ❤️

File: packages/core-js/utils/ast.js (hunk 10 above)
Comment (leomassazza) [reply packages/core-js/utils/ast.js:110]: is checking **contract.**name before ( `>` and `<` ) and later **struct.**name

File: packages/core-js/utils/comparator.js (hunk 3 above)
Comment (leomassazza) [reply packages/core-js/utils/comparator.js:1]: done

File: packages/deployer/internal/storage-ast-validator.js (hunk 9 above)
Comment (leomassazza) [reply packages/deployer/internal/storage-ast-validator.js:104]: The comparator is getting it right, but there's something failing when we get the current artifacts. This works the second time you run the deploy script.

I created a ticket for following up this issue.

File: packages/core-js/test/utils/comparator.spec.js (hunk 8 above)
Comment (leomassazza) [reply packages/core-js/test/utils/comparator.spec.js:78]: done

File: packages/deployer/internal/storage-ast-validator.js (hunk 9 above)
Comment (leomassazza) [reply packages/deployer/internal/storage-ast-validator.js:104]: Additionally added a test to cover that situation in the comparator tests.

**Pull request merged**